common/rfb/H264LibavDecoderContext.cxx

```diff
@@ -0,0 +1,212 @@
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
+ * USA.
+ */
+
+
+extern "C" {
+#include <libavutil/imgutils.h>
+#include <libavcodec/version.h>
+}
+#if LIBAVCODEC_VERSION_MAJOR > 57 || LIBAVCODEC_VERSION_MAJOR == 57 && LIBAVCODEC_VERSION_MINOR >= 37
+#define FFMPEG_DECODE_VIDEO2_DEPRECATED
+#endif
+#if LIBAVCODEC_VERSION_MAJOR >= 58
+#define FFMPEG_INIT_PACKET_DEPRECATED
+#endif
+
+#include <rfb/LogWriter.h>
+#include <rfb/PixelBuffer.h>
+#include <rfb/H264LibavDecoderContext.h>
+
+using namespace rfb;
+
+static LogWriter vlog("H264LibavDecoderContext");
+
+bool H264LibavDecoderContext::initCodec() {
+  sws = NULL;
+  swsBuffer = NULL;
+  h264AlignedBuffer = NULL;
+  h264AlignedCapacity = 0;
+
+  AVCodec *codec = avcodec_find_decoder(AV_CODEC_ID_H264);
+  if (!codec)
+  {
+    vlog.error("Codec not found");
+    return false;
+  }
+
+  parser = av_parser_init(codec->id);
+  if (!parser)
+  {
+    vlog.error("Could not create H264 parser");
+    return false;
+  }
+
+  avctx = avcodec_alloc_context3(codec);
+  if (!avctx)
+  {
+    vlog.error("Could not allocate video codec context");
+    return false;
+  }
+
+  frame = av_frame_alloc();
+  if (!frame)
+  {
+    vlog.error("Could not allocate video frame");
+    return false;
+  }
+
+  if (avcodec_open2(avctx, codec, NULL) < 0)
+  {
+    vlog.error("Could not open codec");
+    return false;
+  }
+
+  int numBytes = av_image_get_buffer_size(AV_PIX_FMT_RGB32, rect.width(), rect.height(), 1);
+  swsBuffer = new uint8_t[numBytes];
+
+  initialized = true;
+  return true;
+}
+
+void H264LibavDecoderContext::freeCodec() {
+  if (!initialized)
+    return;
+  av_parser_close(parser);
+  avcodec_free_context(&avctx);
+  av_frame_free(&frame);
+  delete[] swsBuffer;
+  free(h264AlignedBuffer);
+  initialized = false;
+}
+
+rdr::U8* H264LibavDecoderContext::validateH264BufferLength(rdr::U8* buffer, rdr::U32 len)
+{
+  rdr::U32 reserve_len = len + len % AV_INPUT_BUFFER_PADDING_SIZE;
+  if (len == reserve_len)
+    return buffer;
+
+  if (!h264AlignedBuffer || reserve_len > h264AlignedCapacity)
+  {
+    h264AlignedBuffer = (rdr::U8*)realloc(h264AlignedBuffer, reserve_len);
+    h264AlignedCapacity = reserve_len;
+  }
+
+  memcpy(h264AlignedBuffer, buffer, len);
+  memset(h264AlignedBuffer + len, 0, h264AlignedCapacity - len);
+  return h264AlignedBuffer;
+}
+
+void H264LibavDecoderContext::decode(rdr::U8* h264_buffer, rdr::U32 len, rdr::U32 flags, ModifiablePixelBuffer* pb) {
+  h264_buffer = validateH264BufferLength(h264_buffer, len);
+
+#ifdef FFMPEG_INIT_PACKET_DEPRECATED
+  AVPacket *packet = av_packet_alloc();
+#else
+  AVPacket *packet = new AVPacket();
+  av_init_packet(packet);
+#endif
+
+  int ret;
+  int frames_received = 0;
+  while (len)
+  {
+    ret = av_parser_parse2(parser, avctx, &packet->data, &packet->size, h264_buffer, len, AV_NOPTS_VALUE, AV_NOPTS_VALUE, 0);
+    if (ret < 0)
+    {
+      vlog.error("Error while parsing");
+      break;
```

Comment:
But H264 itself isn't reliable. We can use different coders and decoder which can met or not met together's requirements or can operate in different modes. I would't drop connection here if we have a stream

Reply:
Video codecs tolerate data loss, but I doubt a working encoder will actively cause missing data, so I would still say this indicates a bug somewhere. But if there is a synchronisation problem on the RFB level it will likely be caught elsewhere. So this can stay if you would prefer to keep it.

Reply:
In this case connection shouldn't be dropped definitely. Because it can be a bug/throttling issue on coding side (like in our case the coder is a tiny device with small RISC CPU). So parsing can fail from time to time. It's not an issue on the whole connection's scale. Also I can remove logging here. But I find them handy if you are trying to debug sender issues. This error in real logs shouldn't appear on stable stream and if it appears it indicates a problem but not here, most likely somewhere else. I would keep it as it is if you don't mind.

Reply:
Sure, it can stay. We can always revisit this issue if we get user reports that this is common issue.